### ISSUE TITLE
Container creation and start are occassionally slow

### ISSUE THREAD

**Issue by abhishekram** [docker/for-linux#121]
**Description**
We are doing a performance test where we create and start 500 containers on a [Microsoft Azure D4_v2 Linux VM](https://azure.microsoft.com/en-us/pricing/details/virtual-machines/linux) and measure the time it takes. 

Most of the time, the 500 containers would be created and started in 60 to 75 seconds. However, occasionally we would see it take 100 to 120 seconds.

**Steps to reproduce the issue:**
Invoke the Docker REST APIs to create and start 500 containers. We maintained no more than 100 concurrent container activations at any given time, where activation is defined as the process of creating *and* starting a single container. In other words, we initially started with 100 simultaneous activations. Then, each time a container was successfully activated, we'd start a new container activation. This is how we maintained 100 concurrent activations at any given time. We stopped when we had completed 500 container activations.

**Describe the results you received:**
Most of the time, the 500 containers would be activated in 60 to 75 seconds. However, occasionally we would see it take 100 to 120 seconds.

**Describe the results you expected:**
We expected more consistency in the results, i.e. we expected it to always take between 60 and 75 seconds. We are interested in understanding why it sometimes takes significantly longer.

**Additional information you deem important (e.g. issue happens only occasionally):**
I have attached an Excel file that shows:
* Request processing time for each create and start request, for both slow nodes and fast nodes.
* Charts showing distribution of create and start request processing times, for both slow nodes and fast nodes.
[DockerPerfVariance.xlsx](https://github.com/docker/for-linux/files/1361026/DockerPerfVariance.xlsx)

I have also attached Docker debug level logs for a fast run and a slow run. 
[FastRun-DockerLogs.txt](https://github.com/docker/for-linux/files/1361027/FastRun-DockerLogs.txt)
[SlowRun-DockerLogs.txt](https://github.com/docker/for-linux/files/1361028/SlowRun-DockerLogs.txt)

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 17.09.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 11
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-24-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 27.48GiB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 21
 Goroutines: 30
 System Time: 2017-10-05T21:57:50.029286482Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Microsoft Azure D4_v2 Linux VM

**Comment by thaJeztah** [comment]
@tonistiigi any thoughts on getting more data to localize the bottleneck? Would the prometheus metrics provide insight?

**Comment by abhishekram** [comment]
Hi, any ideas on what might be going on?

**Comment by tonistiigi** [comment]
If I need to guess I would say that this is caused by the rate limiting in containerd (0.2) but I have no data to prove it.

The logs do not contain the containerd request time but the slow part is in between mounting the container and receiving back the call from network hook from containerd.

cc @mlaventure

**Comment by abhishekram** [comment]
Any enhanced logging we can turn on to get more information? Or any other technique?

**Comment by shuaichang** [comment]
@tonistiigi could you point me to the code line where the rate limiting in containerd is set? When you see 0.2 is that the containerd version or a value for the rate limit?

**Comment by tonistiigi** [comment]
@shuaichang Yes, containerd v0.2 https://github.com/containerd/containerd/blob/v0.2.x/supervisor/supervisor.go#L23
